core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,586 @@
+    decode = decodeWord64
+
+instance Bi Int where
+    encode = encodeInt
+    decode = decodeInt
+
+instance Bi Float where
+    encode = encodeFloat
+    decode = decodeFloat
+
+instance Bi Int32 where
+    encode = encodeInt32
+    decode = decodeInt32
+
+instance Bi Int64 where
+    encode = encodeInt64
+    decode = decodeInt64
+
+instance Bi Nano where
+    encode (MkFixed resolution) = encodeInteger resolution
+    decode = MkFixed <$> decodeInteger
+
+instance Bi Void where
+    decode = fail "instance Bi Void: you shouldn't try to deserialize Void"
+    encode = error "instance Bi Void: you shouldn't try to serialize Void"
+
+----------------------------------------------------------------------------
+-- Tagged
+----------------------------------------------------------------------------
+
+instance Bi a => Bi (Tagged s a) where
+    encode (Tagged a) = encode a
+    decode = Tagged <$> decode
+
+----------------------------------------------------------------------------
+-- Containers
+----------------------------------------------------------------------------
+
+instance (Bi a, Bi b) => Bi (a,b) where
+    encode (a,b) = encodeListLen 2
+                <> encode a
+                <> encode b
+    decode = do decodeListLenOf 2
+                !x <- decode
+                !y <- decode
+                return (x, y)
+
+instance (Bi a, Bi b, Bi c) => Bi (a,b,c) where
+    encode (a,b,c) = encodeListLen 3
+                  <> encode a
+                  <> encode b
+                  <> encode c
+
+    decode = do decodeListLenOf 3
+                !x <- decode
+                !y <- decode
+                !z <- decode
+                return (x, y, z)
+
+instance (Bi a, Bi b, Bi c, Bi d) => Bi (a,b,c,d) where
+    encode (a,b,c,d) = encodeListLen 4
+                    <> encode a
+                    <> encode b
+                    <> encode c
+                    <> encode d
+
+    decode = do decodeListLenOf 4
+                !a <- decode
+                !b <- decode
+                !c <- decode
+                !d <- decode
+                return (a, b, c, d)
+
+instance Bi BS.ByteString where
+    encode = encodeBytes
+    decode = decodeBytes
+
+instance Bi Text.Text where
+    encode = encodeString
+    decode = decodeString
+
+instance Bi BS.Lazy.ByteString where
+    encode = encode . BS.Lazy.toStrict
+    decode = BS.Lazy.fromStrict <$> decode
+
+instance Bi a => Bi [a] where
+    encode = encodeList
+    decode = decodeList
+
+instance (Bi a, Bi b) => Bi (Either a b) where
+    encode (Left  x) = encodeListLen 2 <> encodeWord 0 <> encode x
+    encode (Right x) = encodeListLen 2 <> encodeWord 1 <> encode x
+
+    decode = do decodeListLenOf 2
+                t <- decodeWord
+                case t of
+                  0 -> do !x <- decode
+                          return (Left x)
+                  1 -> do !x <- decode
+                          return (Right x)
+                  _ -> fail "unknown tag"
```

Comment:
I think `unknown tag` messages should include the type name to make debugging easier in case something goes wrong. Especially because we don't have `label` (or analogous) function (AFAIU).